core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
 
   private def abortTimedOutTransactions(): Unit = {
     txnManager.timedOutTransactions().foreach { txnIdAndPidEpoch =>
-      handleEndTransaction(txnIdAndPidEpoch.transactionalId,
-        txnIdAndPidEpoch.producerId,
-        txnIdAndPidEpoch.producerEpoch,
-        TransactionResult.ABORT,
-        (error: Errors) => error match {
-          case Errors.NONE =>
-            info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
-          case Errors.INVALID_PRODUCER_ID_MAPPING |
-               Errors.INVALID_PRODUCER_EPOCH |
-               Errors.CONCURRENT_TRANSACTIONS =>
-            debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
-          case e =>
-            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
-        })
-    }
+      val coordinatorEpochAndMetadata = txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
+        case None =>
+          fatal(s"Could not find transaction metadata when trying to timeout transaction with transactionalId " +
```

Comment:
Should this really be logged as a fatal error? I think it may happen that a leader migration happens between the `timedOutTransactions` call and the `foreach` loop here.

In the old code we treat such errors as a debug level logging only:

```
(error: Errors) => error match {
          case Errors.NONE =>
            info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
          case Errors.INVALID_PRODUCER_ID_MAPPING |
               Errors.INVALID_PRODUCER_EPOCH |
               Errors.CONCURRENT_TRANSACTIONS =>
            debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
          case e =>
            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
        })
```

Reply:
The problem here is that this is happening in a background thread. If we are trying to abort a transaction but fail because the internal metadata is inconsistent, that is a severe error. And there is no way to report this back to the user apart from the logs. 

What would happen if we hit this case (which should never happen because it means internal state is inconsistent) is that transactions would never be expired, which means that read committed consumers would never make progress. If that happens, we need clues in the logs to inform us of what happens.

I was debating whether this should be at an error level, and I think that is fine as well. But it should certainly be no lower than error IMO.

Reply:
I changed this to an error log level in the latest update.

Reply:
> If we are trying to abort a transaction but fail because the internal metadata is inconsistent, that is a severe error.

`timedOutTransactions` only made a pass over the metadata and generate a candidate list of transactions that can be expired, and then in `foreach` we iterate over the list and tries to end the transaction as aborted. So if it fails to end it that is fine since the thread will add it to the candidate list again in the next call as long as the metadata is still in the cache, right?